site/src/job_queue.rs

```diff
@@ -0,0 +1,57 @@
+use std::{str::FromStr, sync::Arc};
+
+use crate::load::SiteCtxt;
+use chrono::Utc;
+use database::{BenchmarkRequest, BenchmarkRequestStatus};
+use parking_lot::RwLock;
+use tokio::time::{self, Duration};
+
+/// Store the latest master commits or do nothing if all of them are
+/// already in the database
+async fn enqueue_master_commits(ctxt: &Arc<SiteCtxt>) {
+    let conn = ctxt.conn().await;
+    let master_commits = &ctxt.get_master_commits().commits;
+    // TODO; delete at some point in the future
+    let cutoff: chrono::DateTime<Utc> =
+        chrono::DateTime::from_str("2025-06-01T00:00:00.000Z").unwrap();
+
+    for master_commit in master_commits {
+        // We don't want to add masses of obsolete data
+        if master_commit.time >= cutoff {
+            let pr = master_commit.pr.unwrap_or(0);
+            let benchmark = BenchmarkRequest::create_master(
+                &master_commit.sha,
+                &master_commit.parent_sha,
+                pr,
+                master_commit.time,
+                BenchmarkRequestStatus::WaitingForParent,
+                "",
+                "",
+            );
+            conn.insert_benchmark_request(&benchmark).await;
+        }
+    }
+}
+
+/// For queueing jobs, add the jobs you want to queue to this function
+async fn cron_enqueue_jobs(site_ctxt: &Arc<SiteCtxt>) {
+    // Put the master commits into the `benchmark_requests` queue
+    enqueue_master_commits(site_ctxt).await;
+}
+
+/// Entry point for the cron
+pub async fn cron_main(site_ctxt: Arc<RwLock<Option<Arc<SiteCtxt>>>>, seconds: u64) {
+    let ctxt = site_ctxt.clone();
+    let mut interval = time::interval(Duration::from_secs(seconds));
+
+    if let Some(ctxt_clone) = {
```

Comment:
This condition needs to be inside the loop. First we should tick the interval, than check if the ctxt is filled, and then run the `cron_enqueue_jobs` function. Both because the context is missing when `cron_main` is called (so the cron doesn't run at all), and also because if we took the context here, we would be using an old context for the whole duration of the cron job.

Reply:
Ah! I think this commit should handle that; a145daa6abb53ff3ac22631c896130509311fa21 

I `.clone()` outside of the loop which I think makes sense to increment the reference count. Then do as you say inside the loop and have moved the `interval.tick()` to be the first call to be invoked in the loop

Reply:
The clone is not necessary as you already hold a value of the `Arc` from the parameter, but it doesn't really matter.